### ISSUE TITLE
Tests for Lepton number, Baryon number and a function for R-parity

### ISSUE THREAD

**Issue by amanmdesai** [scikit-hep/particle#498]
As per discussion in #494 with @eduardo-rodrigues 

- [ ] To write tests for lepton number and baryon number 
- [ ] Add a function to extract r-parity of susy particles

c.c.  @eduardo-rodrigues @redeboer

**Comment by eduardo-rodrigues** [comment]
I was not originally thinking about that, as it would require several new quantum numbers to be added to the code here in Particle. That is not to say that it is totally wild. But it would require some (more) work. What do you think?

In the first instance one would add lepton and baryon numbers for the particles, and would take it from there.

Kindly comment and help out :-).

**Comment by eduardo-rodrigues** [comment]
Hello @redeboer, you linked above to code from you. Is this something you would like to port to Particle? It could be very neat. It's not a tiny piece of work. I thought I would ask anyway :-).

**Comment by redeboer** [comment]
Yes this could be ported, though parts have to be rewritten for `particle`'s implementation.
I'm tagging @grayson-helmholz, who is currently reworking QRules. Maybe something you could look into? On the road myself the next few weeks

**Comment by eduardo-rodrigues** [comment]
Hi. Sounds cool. Let's keep in touch.

And happy trips to you :)!

**Comment by eduardo-rodrigues** [comment]
Hello @redeboer, @grayson-helmholz, @amanmdesai. I hope you're doing well!

Listen, I was checking on open issues. The enhancements proposed above seem like something that could be done stepwise, for example adding lepton quantum numbers at first, then ... Is any one of you willing to take this up?

**Comment by rakshithrajs** [comment]
Hi @eduardo-rodrigues , @redeboer 
I am looking to make my first contribution to scikit hep and this issue caught my eye. The stepwise approach Eduardo mentioned sounds like a good way for me to get started without biting off more than I can handle.
Would it be okay if I take the lepton number part first? 
My plan would be to add it as a property on Particle and write tests for the obvious cases. If that goes well I can pick up baryon number and the R parity function in follow-up PRs.
Two things I wasnt sure about from poking around the repo:

where the property should live alongside the PDG ID stuff, or somewhere separate?
whether there s a particular testing pattern I should follow for quantum number tests

I'll hold off until you confirm the scope is reasonable. Thanks!

**Comment by eduardo-rodrigues** [comment]
Hello @rakshithrajs, and thank you for the interest.

Yes, what you say sounds good to me. Better start small and proceed in steps, to ensure we're good everywhere. I hope @redeboer will be able to comment and review as well, since you will effectively be importing a lot from him and collaborators.

The PDGID class does not have all the info encored for all quantum numbers, so you will have to implement the internal quantum numbers on the Particle class, which makes sense. This being said, there are cross-checks with a particle's PDGID that can be made, see for example https://github.com/ComPWA/qrules/blob/0.10.0/src/qrules/particle.py#L214 for lepton numbers. We can look at self-consistency checks along the way; these are very important as "physics-related tests".

I hope this also answers your second question. As for unit tests, just follow the sort of what we have been testing. Seems fine. Unless you have some other ideas?

I look forward to seeing your contributions. Thank you.

**Comment by redeboer** [comment]
I'm a bit lost on the scope of this issue, also because I don't know the design ideas behind the classes of the `particle` package. Is this issue just about implementing `baryon_number` and `lepton_number` and a function like `get_r_parity()`? Is it about extending the test suite to check for consistency of quantum numbers (and catch potential bugs in the databases)? Or is this about making the PDGID class also work on the level of quantum numbers as well as with quark content?

**Comment by eduardo-rodrigues** [comment]
Hi @redeboer, I'm afraid it is all of that :D! (And yes, some of the quantum numbers can work at the "level" of the PDGID class, meaning the PDG ID can itself determine certain quantum numbers, effectively encoding it.)

**Comment by redeboer** [comment]
#760 is a first attempt, just focusing on those quantum numbers. I was thinking, if you want to extend the class with more quantum numbers (e.g. hypercharge so that you can check for Gell-Mann-Nishijima), it may be worth offering that under a 'sub-attribute', like `p.qn.baryon_number`. Not sure what a good name would be for such namespace and if that fits the current design.

**Comment by eduardo-rodrigues** [comment]
Ah, that was fast! I won't even have time today to check it out properly ...

Hmm, I will have to think about a namespace as the one you are suggesting. So far the design has not been so "complex". If you went such a way, can you provide some motivations?

As you PR does not do everything here, I hope it can serve as a starting point for the full thing for @rakshithrajs.

**Comment by redeboer** [comment]
> Hmm, I will have to think about a namespace as the one you are suggesting. So far the design has not been so "complex". If you went such a way, can you provide some motivations?

It's more of a bad feeling about that large `Particle` class that's growing even more now, especially when implementing even more attributes (b6fe3e97).

The PR now also has a test for checking Gell-Mann–Nishijima consistency of the database. It correctly verifies the rule for 558 of the particles in the database.
